
syntax = "proto2";

package pbview.samples;

option optimize_for = SPEED;
option cc_enable_arenas = true;

enum MyEnum {
    MyEnumVal1 = 0;
    MyEnumVal2 = 1;
    MyEnumVal3 = 2;
}

message MySubMsg {
    required int32  id    = 1;
    required string value = 2;
}

message AllTypes {
    optional double   double_field   = 1; 
    optional float    float_field    = 2;
    optional int32    int32_field    = 3;
    optional int64    int64_field    = 4;
    optional uint32   uint32_field   = 5;
    optional uint64   uint64_field   = 6;
    optional sint32   sint32_field   = 7;
    optional sint64   sint64_field   = 8;
    optional fixed32  fixed32_field  = 9;
    optional fixed64  fixed64_field  = 10;
    optional sfixed32 sfixed32_field = 11;
    optional sfixed64 sfixed64_field = 12;
    optional bool     bool_field     = 13;
    optional string   string_field   = 14;
    optional bytes    bytes_field    = 15;

    optional MyEnum   myenum_field   = 16;
    optional MySubMsg mysubmsg_field = 17; 
}

message AllTypesRepeated {
    repeated double   double_field   = 1; 
    repeated float    float_field    = 2;
    repeated int32    int32_field    = 3;
    repeated int64    int64_field    = 4;
    repeated uint32   uint32_field   = 5;
    repeated uint64   uint64_field   = 6;
    repeated sint32   sint32_field   = 7;
    repeated sint64   sint64_field   = 8;
    repeated fixed32  fixed32_field  = 9;
    repeated fixed64  fixed64_field  = 10;
    repeated sfixed32 sfixed32_field = 11;
    repeated sfixed64 sfixed64_field = 12;
    repeated bool     bool_field     = 13;
    repeated string   string_field   = 14;
    repeated bytes    bytes_field    = 15;

    repeated MyEnum   myenum_field   = 16;
    repeated MySubMsg mysubmsg_field = 17; 
}

message AllTypesRepeatedPacked {
    repeated double   double_field   = 1 [packed=true];
    repeated float    float_field    = 2 [packed=true];
    repeated int32    int32_field    = 3 [packed=true];
    repeated int64    int64_field    = 4 [packed=true];
    repeated uint32   uint32_field   = 5 [packed=true];
    repeated uint64   uint64_field   = 6 [packed=true];
    repeated sint32   sint32_field   = 7 [packed=true];
    repeated sint64   sint64_field   = 8 [packed=true];
    repeated fixed32  fixed32_field  = 9 [packed=true];
    repeated fixed64  fixed64_field  = 10 [packed=true];
    repeated sfixed32 sfixed32_field = 11 [packed=true];
    repeated sfixed64 sfixed64_field = 12 [packed=true];
    repeated bool     bool_field     = 13 [packed=true];
    //repeated string   string_field   = 14 [packed=true];
    //repeated bytes    bytes_field    = 15 [packed=true];

    repeated MyEnum   myenum_field   = 16 [packed=true];
    //repeated MySubMsg mysubmsg_field = 17 [packed=true]; 
}
